tests/test_service.py

```diff
@@ -393,13 +393,19 @@ def testAddAlert(self, mockPost):
         res = self.argus.alerts.add(alert)
         self.assertTrue(isinstance(res, Alert))
         self.assertTrue(hasattr(res, "id"))
+        for method in ['get', 'add', 'update', 'delete']:
+            self.assertTrue(hasattr(res.triggers, method), msg='no alert.triggers.{}()'.format(method))
+            self.assertTrue(hasattr(res.notifications, method), msg='no alert.notifications.{}()'.format(method))
 
     @mock.patch('requests.Session.put', return_value=MockResponse(json.dumps(alert_D), 200))
     def testUpdateAlert(self, mockPut):
-        self.argus.alerts.update(testId, Alert.from_dict(alert_D))
+        res = self.argus.alerts.update(testId, Alert.from_dict(alert_D))
         self.assertTrue(isinstance(self.argus.alerts.get(testId), Alert))
         self.assertEquals(self.argus.alerts.get(testId).to_dict(), alert_D)
         self.assertIn((os.path.join(endpoint, "alerts", str(testId)),), tuple(mockPut.call_args))
+        for method in ['get', 'add', 'update', 'delete']:
```

Comment:
Have you considered simply checking the type of the `triggers` and `notifications` objects to be of type `BaseUpdatableModelServiceClient`?

Reply:
Yes, it looked to me though that the actual end user contract is presence of certain methods etc, and the specific type is more of an implementation detail

Reply:
Agree, especially if that is how it is documented. It looks like the current documentation makes no attempt to classify the service objects, on what type of methods can be called on them. This reminds me that I received an offline patch sometime back that touches this same area and I haven't merged it because I realized that the Alert object, by deviating from being a pure data object, muddled the contract. Rest of the APIs just return an object of the corresponding Model object, but this API deviates in the interest of aligning with the original web endpoints. At that time I left it as is, as I was hoping that the API would evolve in due course based on the usage than designing it puristically. Perhaps this is a good time to ponder on it with you. The two quick options are:
- continue to do the current way, i.e., attach service objects to the `Alert`
- keep the `Alert` object as pure data only object (sort of like POPO - Plain Old Python Object) and give `alert-triggers` and `alert-notifications` services by other means.

In either case,  we would need to better document what different services are of what type (readonly, updatable etc.).

Any thoughts?

Reply:
I'll take the liberty of answering the question as if there were two there :-)

*1) What to do now?*

I believe this or similar patch should be applied to make the behavior of `update` consistent with `get` and `add`, regardless of whether the behavior changes in the future.

*2) Pros and cons of the current approach*

The current state is somewhat reflective of DDD (-like) approach:
- `ArgusServiceClient` is a service which 
- `AlertsServiceClient` along with few others are repositories
- `Alert` is an aggregate root
- `Trigger` and `Notification` are its parts

Speaking for it are:
1. its relative simplicity where `Triggers` and `Notifications` perceived as parts of the `alert` have collections/micro-repositories, `triggers` and `notifications`, directly exposed on the alert itself;
2. almost "direct" mapping to and from the Argus Web Service API, potentially making it easier to make the related design decisions, and easier for end users to figure out how to use the Python client if they are familiar with the Web API
  - `VERB /alerts/{alertId}/triggers` <> `alerts.get(alertId).triggers.method(...)` and
  - `VERB /alerts/{alertId}/notifications` <> `alerts.get(alertId).notifications.method(...)`
3. at the same time somewhat abstracting away the underlying Web API, i.e. providing an API which isn't necessarily "leaks" the underlying Web Service too much -- if I were designing an `alert` with `triggers` and `notifications`, it's pretty natural for OOP to place the corresponding CRUD ops right there 

Some of the drawbacks are:
1. inconsistent behavior of the newly created `Alert` vs one received from the service
2. inconsistency between creating an alert and updating it, where the `alert.add` creates alerts and notifications at least in the simple case while `alert.update` doesn't do anything with them
3. the `get(alertId)` above is somewhat unfortunate effect of this approach, as even if one doesn't need the `Alert` currently they still have to retrieve it
4. in practice it isn't always easy to directly map Web API to Python, for example, currently some other operations are expressed in a different way, e.g.  `[get|add|delete]_notification_trigger[s]` are exposed on `AlertsServiceClient`, not `Trigger`, `Notification` or the `Alert` itself

I'd say that above drawbacks aren't conceptual but rather something which can be taken care of with a little effort, for example:
1. `add` doesn't have to accept an alert, rather would accept the args to create it, fixing the interaction with "not-yet-born" `Alert`'s by taking them away from the picture
2. would be taken care of by the (1) above
3. can be optimized away if becomes a significant enough issue, but I probably wouldn't bother at this point
4. these are simple enough to make `Alert` operations if conceptual consistency is desired

All in all, I agree with the calculation of choosing the simpler approach, and would proceed with it for now, fixing things as we go, if/until more downsides are discovered.

Reply:
Thank you, you did a lot better in justifying the approach. I will go ahead and merge now.